microsoft/testsuites/nvme/performance.py

```diff
@@ -0,0 +1,99 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+from enum import Enum
+from typing import List
+
+from assertpy import assert_that
+
+from lisa import (
+    Node,
+    TestCaseMetadata,
+    TestSuite,
+    TestSuiteMetadata,
+    notifier,
+    simple_requirement,
+)
+from lisa.environment import Environment
+from lisa.features import Nvme, NvmeSettings
+from lisa.notifier import FIOResult, FIOResultsMessage
+from lisa.tools import Echo, Fio, Lscpu
+
+FIOMODES = Enum(
+    "FIOMODES",
+    [
+        "randread",
+        "randwrite",
+        "read",
+        "write",
+    ],
+)
+
+
+@TestSuiteMetadata(
+    area="nvme",
+    category="performance",
+    description="""
+    This test suite is to validate NVMe disk performance of Linux VM using fio tool.
+    """,
+)
+class nvmeperf(TestSuite):  # noqa
+    TIME_OUT = 3000
+
+    @TestCaseMetadata(
+        description="""
+        This test case uses fio to test NVMe disk performance.
+        """,
+        priority=2,
+        timeout=TIME_OUT,
+        requirement=simple_requirement(
+            supported_features=[NvmeSettings(disk_count=8)],
+        ),
+    )
+    def verify_nvme_performance(self, node: Node, environment: Environment) -> None:
+        nvme = node.features[Nvme]
+        nvme_namespaces = nvme.get_namespaces()
+        assert_that(len(nvme_namespaces)).described_as(
+            "At least 1 NVMe disk for fio testing."
+        ).is_greater_than(0)
+        filename = ":".join(nvme_namespaces)
+        echo = node.tools[Echo]
+        # This will have kernel avoid sending IPI to finish I/O on the issuing CPUs
+        # if they are not on the same NUMA node of completion CPU.
+        # This setting will give a better and more stable IOPS.
+        for nvme_namespace in nvme_namespaces:
+            # /dev/nvme0n1 => nvme0n1
+            disk_name = nvme_namespace.split("/")[-1]
+            echo.write_to_file(
+                "0",
+                node.get_pure_path(f"/sys/block/{disk_name}/queue/rq_affinity"),
+                sudo=True,
+            )
+        cpu = node.tools[Lscpu]
+        start_qdepth = cpu.get_core_count()
+        max_qdepth = start_qdepth * 256
+        numjob = cpu.get_core_count()
+        numjobiterator = 0
+        fio_result_list: List[FIOResult] = []
+        fio = node.tools[Fio]
+        for mode in FIOMODES:
+            qdepth = start_qdepth
+            while qdepth <= max_qdepth:
+                iodepth = int(qdepth / numjob)
+                fio_result = fio.launch(
+                    name=f"iteration{numjobiterator}",
+                    filename=filename,
+                    mode=mode.name,
+                    iodepth=int(iodepth),
+                    numjob=numjob,
+                )
+                fio_result_list.append(fio_result)
```

Comment:
How about deliver multiple result messages, instead of creating a list and parse it later? It can be sent when the message is generated. It doesn't need to wait until all results are produced.

Reply:
table for fio test results is based on the iodepth, we need collect all test results then parse them, insert them one time.

Reply:
what does the "parse" mean? I see the items insert in database one by one. The shared information can be duplicate to each message. That will make the processing simpler.

Reply:
now results are parsed in test case.